File: coordinates/src/main/box-web-mercator.ts
import { PointWebMercator } from "./point-web-mercator"

/**
 * The point (0,0) is in the lower right
 */
export class BoxWebMercator {
    private topLeft: PointWebMercator;
    private bottomRight: PointWebMercator;

    constructor(topLeftX: number = 0, topLeftY: number = 0, bottomRightX: number = 0, bottomRightY: number = 0) {
        this.topLeft = new PointWebMercator(topLeftX, topLeftY);
        this.bottomRight = new PointWebMercator(bottomRightX, bottomRightY);
    }

    public getTopLeft(): PointWebMercator {
        return this.topLeft;
    }

    public getBottomRight(): PointWebMercator {
        return this.bottomRight;
    }

    public getWidth(): number {
        return this.bottomRight.x - this.topLeft.x
    }

    public getHeight(): number {
        return this.topLeft.y - this.bottomRight.y
    }

    /**
     * Returns the intersection of the two boxes.  An real number range is allowed.
     * @param o 
     */
    public intersection(o: BoxWebMercator): BoxWebMercator | null {
        let newBox = new BoxWebMercator();
        newBox.topLeft.x = Math.max(this.topLeft.x, o.topLeft.x);
        newBox.topLeft.y = Math.min(this.topLeft.y, o.topLeft.y);
        newBox.bottomRight.x = Math.min(this.bottomRight.x, o.bottomRight.x);
        newBox.bottomRight.y = Math.max(this.bottomRight.y, o.bottomRight.y);

        if (newBox.bottomRight.x < newBox.topLeft.x || newBox.bottomRight.y > newBox.topLeft.y) {
            return null;
        }

        return newBox;
    }

}

File: vfr-green-app/src/commonMain/kotlin/BoxWebMercator.kt
import kotlin.js.JsName
import kotlin.math.max;
import kotlin.math.min;

class BoxWebMercator(topLeftX: Double = 0.0, topLeftY: Double = 0.0, bottomRightX: Double = 0.0, bottomRightY: Double = 0.0) {
    var topLeft: PointWebMercator
        private set;
    var bottomRight: PointWebMercator
        private set;

    init {
        this.topLeft = PointWebMercator(topLeftX, topLeftY);
        this.bottomRight = PointWebMercator(bottomRightX, bottomRightY);
    }

    public fun getWidth(): Double {
        return this.bottomRight.x - this.topLeft.x
    }

    public fun getHeight(): Double {
        return this.bottomRight.y - this.topLeft.y
    }

    /**
     * Returns the union of the two boxes.  An real number range is allowed.
     * @param o
     */
    @JsName("union")
    public fun union(o: BoxWebMercator): BoxWebMercator? {
        val newBox = BoxWebMercator();
        newBox.topLeft.x = max(this.topLeft.x, o.topLeft.x);
        newBox.topLeft.y = max(this.topLeft.y, o.topLeft.y);
        newBox.bottomRight.x = min(this.bottomRight.x, o.bottomRight.x);
        newBox.bottomRight.y = min(this.bottomRight.y, o.bottomRight.y);

        if (newBox.bottomRight.x < newBox.topLeft.x || newBox.bottomRight.y < newBox.topLeft.y) {
            return null;
        }

        return newBox;
    }
}